pricegraph/src/graph/shortest_paths.rs

```diff
@@ -64,80 +78,40 @@ where
     pub fn path_to(&self, dest: G::NodeId) -> Option<Path<G::NodeId>> {
-        let max_path_len = self.predecessor_store.predecessors.len();
-        let mut path = Vec::with_capacity(max_path_len);
-        let mut current = dest;
-        while current != self.source {
-            assert!(path.len() <= max_path_len, "undetected negative cycle");
-            path.push(current);
-            current = self.predecessor_store.predecessors[self.graph.to_index(current)]?;
-        }
-        path.push(self.source);
-
-        // NOTE: `path` is in reverse order, since it was built by walking the path
-        // backwards, so reverse it and done!
-        path.reverse();
-        Some(Path(path))
+        self.predecessor_store
+            .path_to(self.source, dest, self.graph)
     }
 
-    /// Lists all nodes that can be reached from the source.
+    /// Lists all nodes that can be reached from the source (including the source itself).
     pub fn connected_nodes(&self) -> Vec<G::NodeId> {
-        let mut node_indices: Vec<_> = self
-            .predecessor_store
-            .predecessors
-            .iter()
-            .enumerate()
-            .filter_map(|(i, &pre)| pre.map(|_| self.graph.from_index(i)))
-            .collect();
+        let mut node_indices = self.predecessor_store.connected_nodes(self.graph);
         debug_assert!(!node_indices.contains(&self.source));
         node_indices.push(self.source);
         node_indices
     }
-}
 
-impl<G> ShortestPathGraph<G>
-where
-    G: IntoNodeIdentifiers + IntoEdges + NodeIndexable,
-    G::EdgeWeight: FloatMeasure,
-{
     /// Initializes a shortest path graph that will be later built with the
     /// Bellman-Ford algorithm.
     fn empty(g: G, source: G::NodeId) -> Self {
         let predecessors = vec![None; g.node_bound()];
-        let mut distance = vec![<_>::infinite(); g.node_bound()];
-        distance[g.to_index(source)] = <_>::zero();
-
-        let predecessor_store = PredecessorStore {
-            predecessors,
-            distance,
-        };
+        let mut distances = vec![<_>::infinite(); g.node_bound()];
+        distances[g.to_index(source)] = <_>::zero();
 
         ShortestPathGraph {
             graph: g,
-            predecessor_store,
+            predecessor_store: Box::new(Unbounded::new(predecessors, distances)),
             source,
         }
     }
 
+    fn prepare_next_relaxation_step(&mut self) {
+        self.predecessor_store.prepare_next_relaxation_step();
+    }
+
     /// Checks for negative weight cycle and, if any is found, creates loop in
     /// the predecessor store
     fn mark_cycle(&mut self) -> Option<G::NodeId> {
-        // This is the last step of the Bellman-Ford algorithm. It tries to relax
-        // each node: if a node can be relaxed then a negative cycle exists and is
-        // created in the predecessor store; otherwise no such cycle exists.
```

Comment:
This comment is gone. Not sure if we need to have it just pointing out in case that was done by accident.

Reply:
👍 will put it back